### ISSUE TITLE
ksmbd and io_uring ?

### ISSUE THREAD

**Issue by Andy2244** [cifsd-team/ksmbd#435]
Just wondering whats your opinion on io_uring and liburing, regarding ksmbd?

**Comment by namjaejeon** [comment]
Ah, I have known io_uring was recently applied to samba. I don't know how AIO is useful in smb server yet. Maybe there's a reason to ask this.

**Comment by Andy2244** [comment]
here is what i checked, last link has benchmarks:
https://wiki.samba.org/index.php/Samba_4.12_Features_added/changed#.27io_uring.27_vfs_module
https://lwn.net/Articles/810414/
https://kernel.dk/io_uring.pdf
https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.6-IO-uring-Tests

**Comment by namjaejeon** [comment]
Yes, cifs doesn't even support AIO, and don't know how there is an usage for this. to work AIO, client do also support it together ?

**Comment by namjaejeon** [comment]
ksmbd I/O use workqueue, i.e. I/O is processed in parallel asynchronously.
So I guess io_uring is not needed for ksmbd.

**Comment by Andy2244** [comment]
> ksmbd I/O use workqueue, i.e. I/O is processed in parallel asynchronously.
> So I guess io_uring is not needed for ksmbd.

Thats what samba and many other projects also did, yet if i understand this correctly the overhead to setup/manage the threads is what makes io_uring preferable over older kernel AIO or handcrafted solution's.
As i understand it the more overlapping async i/o operations you have the better io_uring performs, compared to other approaches. From my tests with the new samba4 vfs_uring module on openwrt+decent router+ssd the gains are minimal for normal sequential operations or light 4 AIO OP's on a single file. Yet as Phoronix has shown the gains get way better if you try something link accessing a file backed database with many I/O requests.

So i think for us (openwrt) users its not really important and just a nice extra feature, while for corporate use cases this may solve some specific workloads or even outperform NFSv4 setups.

**Comment by neheb** [comment]
@Andy2244 is ksmbd slower than samba?

**Comment by namjaejeon** [comment]
ksmbd's performance is better than samba(4.7.6), It is even more different in the latest version of Samba.

See: 
 1. https://www.snia.org/sites/default/files/SDC/2019/presentations/SMB/Jeon_Namjae_Lee_Hyunchul_Introduce_New_In-kernel_SMB3_Server_Called_CIFSD.pdf
 2. https://lists.samba.org/archive/samba-technical/2019-September/134200.html

I hope someone other than me to measure and compare samba and cifsd performance. Including io_uring applied version...

**Comment by Neustradamus** [comment]
@Andy2244: What do you think about performance?

**Comment by Andy2244** [comment]
> 
> 
> @Andy2244: What do you think about performance?

Its a bit tricky, but since io_uring is winning over the linux world and shows decent performance increases in every application it is deployed. The main improvements are on thread heavy and multi-segment file loads, which is more a enterprise/heavy NAS use case side of things. I cant say much since ksmbd is already in kernel space and i'm not a linux i/o coder, so cant judge ksmbd+io_uring vs samba+io_uring.

Yet, since there is a stable uringlib for easy integration, its worth a try since it should also have coding/maintenance advantages over the old i/o calls/framework.
